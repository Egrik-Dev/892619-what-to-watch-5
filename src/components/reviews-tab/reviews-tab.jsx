import React from 'react';
import PropTypes from 'prop-types';
import {filmProps} from '../../props/props';
import {fetchFilmComments} from '../../store/action-api';
import {connect} from 'react-redux';
import {dateParse} from '../../utils';

const ReviewsTab = (props) => {
  const {film, fetchComments} = props;

  let [comments, setComments] = React.useState({});
  let [isFetchingComments, setFlag] = React.useState(true);

  React.useEffect(() => {
    fetchComments(film.id)
      .then(({data}) => setComments(data))
      .then(() => setFlag(false));
  }, []);

  if (isFetchingComments) {
    return <div>Loading...</div>;
  }

  return (
    <React.Fragment>
      <div className="movie-card__reviews movie-card__row">
        <div className="movie-card__reviews-col">
          {comments.map((comment) => (
            <div
              className="review"
              key={comment.id}
            >
              <blockquote className="review__quote">
                <p className="review__text">{comment.comment}</p>

                <footer className="review__details">
                  <cite className="review__author">{comment.user.name}</cite>
                  <time className="review__date" dateTime={comment.date.slice(0, 10)}>{dateParse(comment.date)}</time>
                </footer>
              </blockquote>

              <div className="review__rating">{comment.rating}</div>
            </div>
          ))}
        </div>
      </div>
    </React.Fragment>
  );
};

ReviewsTab.propTypes = {
  film: PropTypes.shape(filmProps),
  fetchComments: PropTypes.func.isRequired
};

const mapDispatchToProps = (dispatch) => ({
  fetchComments(id) {
    return dispatch(fetchFilmComments(id));
  }
});

export {ReviewsTab};
export default connect(null, mapDispatchToProps)(ReviewsTab);
