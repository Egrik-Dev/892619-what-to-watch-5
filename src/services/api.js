import axios from 'axios';
import {BACKEND_URL} from '../const';

const TIMEOUT = 5000;

export const createApi = () => {
  const api = axios.create({
    baseURL: BACKEND_URL,
    timeout: TIMEOUT,
    withCredentials: true
  });

  const onSucess = (response) => response;

  const onFail = (err) => {
    throw err;
  };

  api.interceptors.response.use(onSucess, onFail);

  return api;
};
