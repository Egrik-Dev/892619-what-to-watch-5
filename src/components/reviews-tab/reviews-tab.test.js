import React from 'react';
import renderer from 'react-test-renderer';
import {ReviewsTab} from './reviews-tab';
import thunk from 'redux-thunk';
import {createApi} from '../../services/api';
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";

const film = {
  name: `Bronson`,
  description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
  rating: 3.6,
  director: `Nicolas Winding Refn`,
  genre: `Action`,
  released: 2008,
  id: 1,
  posterImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
  backgroundImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
  backgroundColor: `#73B39A`,
  scoresCount: 109661,
  runTime: 92,
  isFavorite: true,
  videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`]
};

const comments = [
  {
    id: 1,
    user: {
      id: 12,
      name: `Isaak`
    },
    rating: 4.6,
    comment: `Poised to be an instant classic`,
    date: `2020-10-29T13:38:44.769Z`
  },
  {
    id: 2,
    user: {
      id: 12,
      name: `Ben`
    },
    rating: 7.6,
    comment: `Something comment Something comment Something comment`,
    date: `2020-11-22T10:33:02.936Z`
  }
];

const api = createApi({onError: () => {}});

describe(`Render connected to store component`, () => {
  const middlewares = [thunk.withExtraArgument(api)];
  const createMockStore = configureStore(middlewares);
  const mockStore = createMockStore();

  it(`Should ReviewsTab render correctly`, () => {
    const fetchComments = jest.fn(() => Promise.resolve());
    const tree = renderer
      .create(
          <Provider store={mockStore}>
            <ReviewsTab
              film={film}
              fetchCommentsAction={fetchComments}
              isLoadingProp={false}
              commentsProp={comments}
            />
          </Provider>
      );

    fetchComments();
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
