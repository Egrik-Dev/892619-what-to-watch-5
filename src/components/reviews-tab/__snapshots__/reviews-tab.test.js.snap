// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should ReviewsTab render correctly 1`] = `
<div>
  Loading...
</div>
`;
